pkg/store/consul/session.go

```diff
@@ -211,6 +176,65 @@ type unlocker struct {
 	key string
 }
 
+// TxnUnlocker represents a lock that is held in consul, with convenience
+// functions for operating with the lock
+type TxnUnlocker interface {
+	// UnlockTxn adds consul operations to the transaction within the passed
+	// context to unlock the lock. The caller must call transaction.Commit() with
+	// the passed context for the lock to actually be released
+	UnlockTxn(ctx context.Context) error
+
+	// CheckLockedTxn adds consul operations to the transaction within the passed
+	// context to check that the lock is still held. This is useful for callers
+	// that wish for the operations they are attempting to fail if the lock is no
+	// longer held for any reason
+	CheckLockedTxn(ctx context.Context) error
+}
+
+type txnUnlocker struct {
+	session string
+
+	key string
+}
+
+func (t *txnUnlocker) UnlockTxn(unlockCtx context.Context) error {
+	// check that this session still holds the lock
+	err := transaction.Add(unlockCtx, api.KVTxnOp{
+		Verb:    api.KVCheckSession,
+		Key:     t.key,
+		Session: t.session,
+	})
+	if err != nil {
+		return util.Errorf("could not build key unlocking transaction: %s", err)
+	}
+
+	// delete the lock (which will fail if we're not still holding it
+	// thanks to the check-session operation
```

Comment:
this comment makes assumptions about what other operations are in the transaction. Is that wise?